Metadata Training (#611)
nextflow-io/training: docs/side_quests/metadata.md

- `data`: paths to `.txt` files that contain phrases in different languages

```console title="samplesheet.csv"
id,character,recording
sampleA,squirrel,/workspaces/training/side-quests/metadata/data/bonjour.txt
sampleB,tux,/workspaces/training/side-quests/metadata/data/guten_tag.txt
sampleC,sheep,/workspaces/training/side-quests/metadata/data/hallo.txt
sampleD,turkey,/workspaces/training/side-quests/metadata/data/hello.txt
sampleE,stegosaurus,/workspaces/training/side-quests/metadata/data/hola.txt
sampleF,moose,/workspaces/training/side-quests/metadata/data/salut.txt
sampleG,turtle,/workspaces/training/side-quests/metadata/data/ciao.txt
```

## 1. Read in samplesheet

### 1.1. Read in samplesheet with splitCsv

Let's start by reading in the samplesheet with `splitCsv`. In the main workflow file, you'll see that we've already started the workflow.

```groovy title="main.nf" linenums="1"
workflow  {

    ch_samplesheet = Channel.fromPath("./data/samplesheet.csv")
                            .splitCsv(header: true)

}
```

!!! note

    Throughout this tutorial, we'll use the `ch_` prefix for all channel variables to clearly indicate they are Nextflow channels.

> === "After"

Author: yes for consistency with other trainings

    ```groovy title="main.nf" linenums="2" hl_lines="2-3"
    ch_samplesheet = Channel.fromPath("./data/samplesheet.csv")
                            .splitCsv(header: true)
                            .view()
    ```

=== "Before"

    ```groovy title="main.nf" linenums="2"
    ch_samplesheet = Channel.fromPath("./data/samplesheet.csv")
    ```

> We can use the [`splitCsv` operator](https://www.nextflow.io/docs/latest/operator.html#splitcsv) to split the samplesheet into a channel of maps, where each map represents a row from the CSV file.

Author: https://github.com/nextflow-io/training/commit/41f19f31374dc360bceb7bff7a70f3f411036b87

[[id:sampleG, character:turtle], it]
```

Neat, for each of our samples, we now have a language predicted. You may have noticed something else: we kept the meta data of our samples and associated it with our new piece of information. We achieved this by adding the `meta` map to the output tuple in the process:

```groovy title="main.nf" linenums="12"
output:
      tuple val(meta), stdout
```

This is a useful way to ensure the sample-specific meta information stays connected with any new information that is generated.

### 2.2 Associate the language prediction with the input file

At the moment, our sample files and their language prediction are separated in two different channels: `ch_samplesheet` and `ch_predictions`. But both channels have the same meta information associated with the interesting data points. We can use the meta map to combine our channels back together.

Nextflow includes many methods for combining channels, but in this case the most appropriate operator is [`join`](https://www.nextflow.io/docs/latest/operator.html#join). If you are familiar with SQL, it acts like the `JOIN` operation, where we specify the key to join on and the type of join to perform.

> If we check the [`join`](https://www.nextflow.io/docs/latest/operator.html#join) documentation, we can see that it joins two channels based on a defined item, by default the first item in each tuple. If you don't have the console output still available, let's run the pipeline to check our data structures. If you removed the `view()` operator from the `ch_samplesheet` add it back in:

Reviewer: I agree, this should be a two-step thing:

1) just pass the input file to the output (they've seen that pattern in earlier sections if they've done Genomics)
2) actually, that's a bad plan, because X. What we usually do is keep them separate to keep channel structure simple, yada yada yada

> If we check the [`join`](https://www.nextflow.io/docs/latest/operator.html#join) documentation, we can see that it joins two channels based on a defined item, by default the first item in each tuple. If you don't have the console output still available, let's run the pipeline to check our data structures. If you removed the `view()` operator from the `ch_samplesheet` add it back in:

Author: I was first thinking that we could remove it completely, but we need it to make sure the right prediction ends up with the right file. I have been playing around with it, and there is too much stuff happening between this step and the next process to simply try and fail and try again with join. we might be able to show that channels don't have order with a view but it is not guaranteed

> If we check the [`join`](https://www.nextflow.io/docs/latest/operator.html#join) documentation, we can see that it joins two channels based on a defined item, by default the first item in each tuple. If you don't have the console output still available, let's run the pipeline to check our data structures. If you removed the `view()` operator from the `ch_samplesheet` add it back in:

Author: thought about this more. i think I misunderstood you orginally. I rewrote this now to remove this section entirely and just pass the file through and leave it at that. We can talk about efficient ways to implement this in splitting and grouping I think and just stay focus on meta maps here. 

I assume you meant:

```
      tuple val(meta), path(file), stdout
```

=== "After"

    ```groovy title="main.nf" linenums="20" hl_lines="8 11"
    workflow  {

          ch_samplesheet = Channel.fromPath("./data/samplesheet.csv")
                      .splitCsv(header: true)
                      .map { row ->
                          [ [id:row.id, character:row.character], row.recording ]
                      }

>                       .view()

Reviewer: ```suggestion
                      .view{ "samplesheet item: " + it }
```
We should do this sort of thing if there are going to be multiple view ops active at once.

Author: removed becaus join was removed

    workflow  {

        ch_samplesheet = Channel.fromPath("./data/samplesheet.csv")
                    .splitCsv(header: true)
                    .map { row ->
                        [ [id:row.id, character:row.character], row.recording ]
                    }
                    .view()

        ch_prediction = IDENTIFY_LANGUAGE(ch_samplesheet)
        ch_prediction.view()

    }
    ```

```bash title="View joined channel"
nextflow run main.nf
```

```console title="View joined channel"
[[id:sampleA, character:squirrel], /workspaces/training/side-quests/metadata/data/bonjour.txt, fr]
[[id:sampleB, character:tux], /workspaces/training/side-quests/metadata/data/guten_tag.txt, de]
[[id:sampleC, character:sheep], /workspaces/training/side-quests/metadata/data/hallo.txt, de]
[[id:sampleD, character:turkey], /workspaces/training/side-quests/metadata/data/hello.txt, en]
[[id:sampleE, character:stegosaurus], /workspaces/training/side-quests/metadata/data/hola.txt, es]
[[id:sampleF, character:moose], /workspaces/training/side-quests/metadata/data/salut.txt, fr]
[[id:sampleG, character:turtle], /workspaces/training/side-quests/metadata/data/ciao.txt, it]
```

> It is becoming a bit hard to see, but if you look all the way on the right side, you can see that now each of our language predictions is associated with our input files.

Reviewer: ```suggestion
It is becoming a bit hard to see, but if you look all the way on the right side, you can see that now each of our language predictions is associated with our input files.
Each channel is not composed of 3 items:

1. The meta map (`[id:sampleA, character:squirrel]`)
2. The input file (`/workspaces/training/side-quests/metadata/data/bonjour.txt`)
3. The language identifier (`fr`)
```

Just clarify, rather than highlighting the difficulty in parsing.

[[id:sampleC, character:sheep, lang:de], /workspaces/training/side-quests/metadata/data/hallo.txt]
[[id:sampleD, character:turkey, lang:en], /workspaces/training/side-quests/metadata/data/hello.txt]
[[id:sampleE, character:stegosaurus, lang:es], /workspaces/training/side-quests/metadata/data/hola.txt]
[[id:sampleF, character:moose, lang:fr], /workspaces/training/side-quests/metadata/data/salut.txt]
[[id:sampleG, character:turtle, lang:it], /workspaces/training/side-quests/metadata/data/ciao.txt]
```

Nice, we expanded our meta map with new information we gathered in the pipeline. Let's take a look at what happened here:

After joining our channels, each element looks like this:

```console
[meta, file, lang]  // e.g. [[id:sampleA, character:squirrel], bonjour.txt, fr]
```

The `map` operator takes each channel element and processes it to create a modified version. Inside the closure `{ meta, file, lang -> ... }`, we then take the existing `meta` map, create a new map `[lang:lang]`, and merge both together using `+`, Groovy's way of combining maps.

### 2.4 Assign a language group using a ternary operator

> Alright, now that we have our language predictions, let's use the information to assign them into new groups. In our example data, we have provided data sets that belong either to `germanic` (either English or German) or `romanic` (French, Spanish, Italian) languages.

Reviewer: Definitely romance.

We can use the `map` operator and an [ternary operator](https://groovy-lang.org/operators.html#_ternary_operator) to assign either group. The ternary operator, is a short cut to an if/else clause. It says:

```console title="Ternary"
variable = <condition> ? 'if-the-condition-is-true' : 'Default'
```

and is the same as:

```console title="If/else"
if (<condition>){
  variable = 'if-the-condition-is-true'
} else {
  variable = 'Default'
}
```

=== "After"

    ```groovy title="main.nf" linenums="20" hl_lines="15-18"
    workflow  {

      ch_samplesheet = Channel.fromPath("./data/samplesheet.csv")
                      .splitCsv(header: true)
                      .map { row ->
                          [ [id:row.id, character:row.character], row.recording ]
                      }

      ch_prediction = IDENTIFY_LANGUAGE(ch_samplesheet)

      ch_languages = ch_samplesheet.join(ch_prediction)

>                                   .map { meta, file, lang ->
>                                       [ meta + [lang:lang], file ]
>                                   }
>                                   .map{ meta, file ->
>                                       def lang_group = (meta.lang.equals('de') || meta.lang.equals('en')) ? 'germanic' : 'romanic'
>                                       [ meta + [lang_group:lang_group], file ]
>                                   }

Reviewer: ```suggestion
                                  .map { meta, file, lang ->
                                      def lang_group = (lang in ['de', 'en']) ? 'germanic' : 'romanic'
                                      [ meta + [ lang: lang, lang_group: lang_group ], file ]
                                  }
```

Single map?

>                                   .view()

Author: good point. You are introducing functions later. I am considering just doing a basic if/else. I like ternaries and I think we need to introduce them because they are everywhere, but maybe later

>                                   .view()

Author: https://github.com/nextflow-io/training/commit/7f6f29996692299f0810ea4d81efde0b2aa076cf

[[id:sampleC, character:sheep, lang:de, lang_group:germanic], /workspaces/training/side-quests/metadata/data/hallo.txt]
[[id:sampleD, character:turkey, lang:en, lang_group:germanic], /workspaces/training/side-quests/metadata/data/hello.txt]
[[id:sampleE, character:stegosaurus, lang:es, lang_group:romanic], /workspaces/training/side-quests/metadata/data/hola.txt]
[[id:sampleF, character:moose, lang:fr, lang_group:romanic], /workspaces/training/side-quests/metadata/data/salut.txt]
[[id:sampleG, character:turtle, lang:it, lang_group:romanic], /workspaces/training/side-quests/metadata/data/ciao.txt]
```

Let's understand how this transformation works. The `map` operator takes a closure that processes each element in the channel. Inside the closure, we're using a ternary operator to create a new language group classification.

The ternary expression `(meta.lang.equals('de') || meta.lang.equals('en')) ? 'germanic' : 'romanic'` works like this:

- First, it evaluates the condition before the `?`: checks if the language is either German ('de') or English ('en')
- If the condition is true (language is German or English), it returns 'germanic'

> - If the condition is false (any other language), it returns 'romanic'

Reviewer: I say just scrap this bit. You talk about ternaries above and point to docs, that's sufficient.

> - If the condition is false (any other language), it returns 'romanic'

Author: https://github.com/nextflow-io/training/commit/7f6f29996692299f0810ea4d81efde0b2aa076cf

removed it and instead added a sentence about merging of maps and adding new keys

We store this result in the `lang_group` variable and then add it to our meta map using `meta + [lang_group:lang_group]`. The resulting channel elements maintain their `[meta, file]` structure, but the meta map now includes this new classification. This allows us to group samples by their language family later in the workflow.

### Takeaway

> In this section, you've learned:
> 
> - **Merging on meta maps**: You used `join` to combine two channels based on their meta maps to maintain relationships across processes and channels
> - **Creating custom keys**: You created two new keys in your meta map, adding them with `meta + [new_key:value]`. One based on a computed value from a process, and one based on a condition you set in the `map` operator.
> - **Ternary operator**: You used the ternary operator to determine which language belongs to which group.

Reviewer: ```suggestion
In this section, you've learned how to:

- **Merge on meta maps**: You used `join` to combine two channels based on their meta maps to maintain relationships across processes and channels
- **Create custom keys**: You created two new keys in your meta map, adding them with `meta + [new_key:value]`. One based on a computed value from a process, and one based on a condition you set in the `map` operator.
- **Use a ternary operator**: You used the ternary operator to determine which language belongs to which group.
```

These allow you to associated new and existing meta data with files as you progress through your pipeline.

---

## 3. Filter data based on meta map values

We can use the [`filter` operator](https://www.nextflow.io/docs/latest/operator.html#filter) to filter the data based on a condition. Let's say we only want to process romanic language samples further. We can do this by filtering the data based on the `lang_group` field. Let's create a new channel that only contains romanic languages and `view` it:

=== "After"

    ```groovy title="main.nf" linenums="20" hl_lines="20-23"
    workflow  {

>       ch_samplesheet = Channel.fromPath("./data/samplesheet.csv")

Reviewer: You don't need to repeat such big sections, it just detracts from the message

```

We have successfully filtered the data to only include romanic samples. Let's recap how this works. The `filter` operator takes a closure that is applied to each element in the channel. If the closure returns `true`, the element is included in the output channel. If the closure returns `false`, the element is excluded from the output channel.

In this case, we want to keep only the samples where `meta.lang_group == 'romanic'`. In the closure, we first know that our channel elements are all of shape `[meta, file]` and we can then access the individual keys of the meta map. We then check if `meta.lang_group` is equal to `'romanic'`. If it is, the sample is included in the output channel. If it is not, the sample is excluded from the output channel.

```groovy title="main.nf" linenums="4"
.filter { meta,file -> meta.lang_group == 'romanic' }
```

### Takeaway

In this section, you've learned:

- How to use `filter` to select samples based on metadata

We now have only the romanic language samples left and can process those further. Next we want to make characters say the phrases.

---

## 4. Customize a process with meta map

> Let's let the characters say the phrases that we have passed in. In the [hello-nextflow training](../hello_nextflow/05_hello_containers.md), you already encountered the `cowpy` package, a python implementation of a tool called `cowsay` that generates ASCII art to display arbitrary text inputs in a fun way. We will re-use a process from there.

Reviewer: ```suggestion
Let's let some fun characters say the phrases that we have passed in. In the [hello-nextflow training](../hello_nextflow/05_hello_containers.md), you already encountered the `cowpy` package, a python implementation of a tool called `cowsay` that generates ASCII art to display arbitrary text inputs in a fun way. We will re-use a process from there.
```

'the' implies we know what characters you're referring to, which we don't yet.

Let's customize the characters by changing the `cowpy` command:

=== "After"

    ```groovy title="main.nf" linenums="38" hl_lines="3"
    script:
    """
    cat $input_file | cowpy -c ${meta.character} > cowpy-${input_file}
    """
    ```

=== "Before"

    ```groovy title="main.nf" linenums="38"
    script:
    """
    cat $input_file | cowpy -c "stegosaurus" > cowpy-${input_file}
    """
    ```

Let's run this:

```bash title="Use cowpy"
nextflow run main.nf -resume
```

and take another look at our french phrase:

```console title="fr/cowpy-salut.txt"
 ____________________
/ Salut, ça va?      \
\ à plus dans le bus /
 --------------------
  \
   \   \_\_    _/_/
    \      \__/
           (oo)\_______
           (__)\       )\/\
               ||----w |
               ||     ||
```

This is a subtle difference to other parameters that we have set in the pipelines in previous trainings. A parameter that is passed as part of the `params` object is generally applied to all samples. When a more surgical approache is necessary, using the sample specific information is a good alternative.

### Takeaway

> In this section, you've learned:
> 
> - **Tweaking directives using meta values**: Using meta map values in `publishDir` directives to create dynamic output paths based on sample properties
> 
> - **Tweaking script section based on meta values**: Customizing tool parameters per sample using meta information in the `script` section

Reviewer: ```suggestion
In this section, you've learned how to:

- **Tweak directives using meta values**: Using meta map values in `publishDir` directives to create dynamic output paths based on sample properties

- **Tweak the script section based on meta values**: Customizing tool parameters per sample using meta information in the `script` section
```
